Pharmacogenetics

Bella2347 opened the pull request:
In the Sequencing QC step for KAPA Twist send Solid Tumor samples to the Pharmacogenetics step.

Ovarialcancer normal samples are already routed to an extra step. Wrote a new scrip to work for both ovarialcancer and pharmacogenetics. 

In the pharmacogenetics step a html report is retrived for each sample. The QC in the report is checked and patient information is added. A PDF report is created from the html report and attached to each sample in Clarity. When the step is completed the reports are moved to the RS LIMS folder.

Bella2347 commented:
```
WARNING: Ignored `image-rendering: smooth` at 18:7, invalid value.
WARNING: Ignored `white-space: -moz-pre-wrap` at 184:7, invalid value.
WARNING: Ignored `white-space: -pre-wrap` at 185:7, invalid value.
WARNING: Ignored `white-space: -o-pre-wrap` at 186:7, invalid value.
WARNING: Error: Expected <ident> for declaration name, got hash. at 202:7.
WARNING: Error: Expected <ident> for declaration name, got hash. at 216:7.
```
I get these warning when I create the report (ran from command line)
@ramsainanduri can you fix it?

Bella2347 commented:
[kapaTwist/OvarialNormalPaired_routeToDataAnalysis.py](https://github.com/Clinical-Genomics-Lund/clarity_scripts/pull/57/files#diff-9e629ba6cc8311481d3da307dab191d561767d066309c3b3ccb58d71ec4e9341) has been replaced by [kapaTwist/KAPA_routeSamplesDataAnalysis.py](https://github.com/Clinical-Genomics-Lund/clarity_scripts/pull/57/files#diff-fd85b58d8e52f2308b71a84a3df7ac1c76b4760ce4886bcde47ea5234ede8ae0)
ReQtest test for this: [(KAPA) Sequencing QC - Routing](https://secure.reqtest.com/app/tests/test-case/view/view-detail/4048972/details) and [(KAPA) Sequencing QC](https://secure.reqtest.com/app/tests/test-case/view/view-detail/3815391/details)

Bella2347 commented:
ReQtest test case for pharmacogenetics step [(Pharmacogenetics) Data Analysis](https://secure.reqtest.com/app/tests/test-case/view/view-detail/4048981/details)

ramsainanduri commented:
> ```
> WARNING: Ignored `image-rendering: smooth` at 18:7, invalid value.
> WARNING: Ignored `white-space: -moz-pre-wrap` at 184:7, invalid value.
> WARNING: Ignored `white-space: -pre-wrap` at 185:7, invalid value.
> WARNING: Ignored `white-space: -o-pre-wrap` at 186:7, invalid value.
> WARNING: Error: Expected <ident> for declaration name, got hash. at 202:7.
> WARNING: Error: Expected <ident> for declaration name, got hash. at 216:7.
> ```
> 
> I get these warning when I create the report (ran from command line) @ramsainanduri can you fix it?

I have fixed this upstream from the pipeline

ramsainanduri commented on pharmacogenetics/pharmacogenetics_createReport.py:
Looks good

ramsainanduri reviewed (approved):
LGTM. I have not tested the scripts responsible for moving file to RS LIMS

Bella2347 commented:
Updated with to commits so that it looks for a report named as the solid tumor sample. 

I updated the test-case with two new samples:
- ST-NORMAL # report should be named ST-TUMOR
- ST-MISSING-PAIR # paired name is missing, should give error

ramsainanduri commented:
Updated report path on Lennart

alkc commented on common/COMMON_moveReportToRSlims.py:
```diff
@@ -1,59 +1,65 @@
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "customPackageSMD"))
+from py3functions import setUpArgs, setUpLIMS, set_logger_log_start_script, logger_end_script
+
+RS_LIMS_PAHT = "/mnt/limsrs-clarity/PDF/in/"
```
```suggestion
RS_LIMS_PATH = "/mnt/limsrs-clarity/PDF/in/"
```

typo. remember to change elsewhere.

alkc commented on common/COMMON_moveReportToRSlims.py:
```diff
@@ -1,59 +1,65 @@
+    for action in inart_actions:
+        inart = action["artifact"]
+        # Only move report if sample is marked as complete
+        if not "complete" in action["action"]:
```
What's in `action["action"]` ?

Is it a list or string?

Bella2347 replied:
That are things I also wonder and have to figure out. 
`process.step.actions.next_actions` returns a list of dicts. Each dict looks like this 
`{"artifact" : Artifact, "action" : str}`
Action can only be some predefined values (complete, remove, repeat, nextstep, completerepeat and more I don't know).
I added a comment about this.

alkc commented on common/COMMON_moveReportToRSlims.py:
```diff
@@ -1,59 +1,65 @@
+            logging.info(f"Sample not complete, report not moved, {inart.name}")
+        else:
+            outart = get_outart(outarts, inart)
+            if not outart.files:
+                logging.info(f"No report for {outart.name}")
+            else:
+                # Get the name of the file in Clarity LIMS
+                filename = outart.files[0].original_location.split("/")[-1]
+                # Create the path to the RS LIMS folder
+                path = RS_LIMS_PAHT + filename
+                # Get the file as a readcontent object
+                read_content = lims.get_file_contents(uri=outart.files[0].uri)
+
+                project = outart.samples[0].project
+                if project.udf.get("Project Classification", "") == "Laborativ - Klinisk":
+                    # Write binary report
+                    with open(path, "wb") as f:
+                        f.write(read_content.read())
+                    logging.info(f"Report moved to RS Lims, {filename}")
+                else:
+                    logging.info(f"Test sample, report not moved, {outart.name}")
```
```suggestion
            if not outart.files:
                logging.info(f"No report for {outart.name}")
                continue

            # Get the name of the file in Clarity LIMS
            filename = outart.files[0].original_location.split("/")[-1]
            # Create the path to the RS LIMS folder
            path = RS_LIMS_PAHT + filename
            # Get the file as a readcontent object
            read_content = lims.get_file_contents(uri=outart.files[0].uri)

            project = outart.samples[0].project
            if project.udf.get("Project Classification", "") == "Laborativ - Klinisk":
                # Write binary report
                with open(path, "wb") as f:
                    f.write(read_content.read())
                logging.info(f"Report moved to RS Lims, {filename}")
            else:
                logging.info(f"Test sample, report not moved, {outart.name}")
```

You can just add a `continue` after L43  to advance the for loop and drop the first else-block.

alkc commented on pharmacogenetics/pharmacogenetics_createReport.py:
```diff
@@ -0,0 +1,167 @@
+    logging.info("Uploading file to Clarity LIMS")
+    lims.upload_new_file(artifact, tmp_pdf_path)
+
+    remove_tmp(dir_path, tmp_html_path, tmp_pdf_path)
```
Look into tempfile stdlib module for python: 

https://docs.python.org/3/library/tempfile.html

It takes care of automatic creation and removal of tmp files for you.

If the removal of the tmp file is important (for example to avoid unnecessary storage of personal data) and if you don't want to switch to the tempfile module right now, then consider wrapping the code between L123-L131 in a try-except-finally block, where the `remove_tmp(dir_path, tmp_html_path, tmp_pdf_path)` is added in the `finally:` block to ensure the tmp file is removed even if the code in `try:` block fails.

Bella2347 replied:
I did not know about that one!

alkc commented on common/COMMON_moveReportToRSlims.py:
```diff
@@ -1,59 +1,65 @@
-from genologics.entities import *
-import sys
-import getopt
+from genologics.entities import Process
```
👍

alkc commented on common/COMMON_moveReportToRSlims.py:
```diff
@@ -1,59 +1,65 @@
+RS_LIMS_PAHT = "/mnt/limsrs-clarity/PDF/in/"
+BUTTON = False
+
+def get_outart(outarts, inart):
```
Consider adding type hints to yr functions

alkc commented on common/COMMON_moveReportToRSlims.py:
```diff
@@ -1,59 +1,65 @@
+                # Get the name of the file in Clarity LIMS
+                filename = outart.files[0].original_location.split("/")[-1]
+                # Create the path to the RS LIMS folder
+                path = RS_LIMS_PAHT + filename
```
```suggestion
                path = RS_LIMS_PATH + filename
```

alkc commented on pharmacogenetics/pharmacogenetics_createReport.py:
```diff
@@ -0,0 +1,167 @@
+Script Classification: Critical - One must be on the server to create the pdf report.
+"""
+from genologics.lims import *
+from genologics.entities import *
```
Try to avoid star imports.

alkc commented on pharmacogenetics/pharmacogenetics_createReport.py:
```diff
@@ -0,0 +1,167 @@
+REPORT_PATH_LENNART = "/mnt/access/solid_hg38/pgx/report"
+TMP_PATH = "/opt/gls/clarity/customextensions/tmp_files"
+
+def retrive_html_report(sample):
```
```suggestion
def retrieve_html_report(sample):
```

typo

alkc commented on pharmacogenetics/pharmacogenetics_createReport.py:
```diff
@@ -0,0 +1,167 @@
+    return file_content
+
+def check_QC(file_content, sample_name):
+    """Check QC values in file and complie result for failed QC values.
```
```suggestion
    """Check QC values in file and compile result for failed QC values.
```

alkc reviewed (changes requested):
Some comments. Looks good otherwise!

alkc commented on pharmacogenetics/pharmacogenetics_createReport.py:
```diff
@@ -103,34 +105,22 @@ def add_patient_information(file_content, sample):
-    logging.info("Writing html")
-    with open(tmp_html_path, "w") as tmp_file:
-        tmp_file.write(file_content)
+    with TemporaryDirectory() as tmpdir:
```
Nice! 👍

alkc commented on common/COMMON_moveReportToRSlims.py:
```diff
@@ -34,29 +35,33 @@ def main():
-                else:
-                    logging.info(f"Test sample, report not moved, {outart.name}")
+        # Get the name of the file in Clarity LIMS
+        filename = outart.files[0].original_location.split("/")[-1]
```
```suggestion
        filename = os.path.basename(outart.files[0].original_location)
```

since`os` is already imported in this module

alkc reviewed (approved):
Looks good!

Bella2347 commented:
Test run on [ReQtest](https://secure.reqtest.com/app/tests/test-run/view/view-test-run-details/752227)
